src/GafferScene/Instancer.cpp

```diff
@@ -2356,3 +2525,319 @@ void Instancer::PrototypeScope::setPrototype( const EngineData *engine, const Sc
+	}();
+
+	std::vector< ConstEngineDataPtr > engines( sampleTimes.size() );
+	for( unsigned int i = 0; i < sampleTimes.size(); i++ )
+	{
+		scope.setFrame( sampleTimes[i] );
+		engines[i] = m_instancer->engine( enginePath, scope.context() );
+	}
+
+	scope.setFrame( onFrameTime );
+
+	const std::vector<int> &pointIndicesForPrototype = engines[0]->pointIndicesForPrototype( root().back() );
+	const ScenePlug::ScenePath *prototypeRoot = engines[0]->prototypeRoot( root().back() );
+
+	const ScenePlug *prototypesPlug = m_instancer->prototypesPlug();
+
+	const IECore::MurmurHash outerCapsuleHash = [this](){
+		IECore::MurmurHash result;
+		hash( result );
+		return result;
+	}();
+
+	const bool hasAttributes = engines[0]->numInstanceAttributes() > 0;
+
+	// If constantPrototype is set, then every instance will use the same prototype.
+	ConstPrototypePtr constantPrototype;
+	if( !engines[0]->hasContextVariables() )
+	{
+		constantPrototype = new Prototype(
+			prototypesPlug, prototypeRoot, sampleTimes, outerCapsuleHash, renderOptions,
+			scope.context(), renderer,
+			// If we don't have instance attributes, we can prepare renderer attributes ahead of time
+			!hasAttributes
+		);
+	}
+
+	// If constantPrototype is not set, we will put prototypes in this cache whenever we first encounter
+	// a prototype using a given context.
+	IECorePreview::LRUCache<IECore::MurmurHash, ConstPrototypePtr, IECorePreview::LRUCachePolicy::Parallel, PrototypeCacheGetterKey> prototypeCache(
+		[
+			&prototypesPlug, &prototypeRoot, &sampleTimes, &outerCapsuleHash, &renderOptions,
+			&renderer, &hasAttributes
+		]
+		( const PrototypeCacheGetterKey &key, size_t &cost, const IECore::Canceller *canceller ) -> ConstPrototypePtr
+		{
+			cost = 1;
+			return new Prototype(
+				prototypesPlug, prototypeRoot, sampleTimes, outerCapsuleHash, renderOptions,
+				key.context, renderer,
+				// If we don't have instance attributes, we can prepare renderer attributes ahead of time
+				!hasAttributes
+			);
+		},
+		pointIndicesForPrototype.size() // Never evict, even if prototypes are all unique
+	);
+
+	// We've found problems with performance when running too many iterations in parallel, which appear
+	// to be related with hitting AiNode too hard in parallel ( perhaps related to threads spread between
+	// separate processors ). To partially solve this, we set the grain size so that we shouldn't use more
+	// than 32 threads, which appears to help some in testing.
+	size_t grainSize = std::max( (size_t)1, pointIndicesForPrototype.size() / 32 );
+	task_group_context taskGroupContext( task_group_context::isolated );
+	tbb::parallel_for( tbb::blocked_range<size_t>( 0, pointIndicesForPrototype.size(), grainSize ), [&]( const tbb::blocked_range<size_t> &r )
+		{
+			vector<M44f> pointTransforms( sampleTimes.size() );
+			std::string name;
+			IECoreScenePreview::Renderer::AttributesInterfacePtr attribsStorage;
+
+			Context::EditableScope prototypeScope( scope.context() );
```

Comment:
I _think_ we want to pass the full `ThreadState` in here, rather than just `scope.context()`, so that any active `Monitors` get to see the work we're doing.